specs/hacspec-lib/src/bit_vector.rs

```diff
@@ -0,0 +1,208 @@
+        if byte_index >= self.bytes.len() {
+            return None;
+        }
+
+        let out = self.bytes.bit(self.bit);
+        self.bit += 1;
+
+        Some(out)
+    }
+}
+
+/// Iterator over bits with a windows size.
+///
+/// Each element of the iterator are `window_size` bits.
+pub struct BitsSlicedIter<'a> {
+    bytes: &'a [u8],
+    bit: usize,
+    window_size: usize,
+}
+
+impl Iterator for BitsSlicedIter<'_> {
+    type Item = u16;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // let byte_index = self.bit / 8;
+        let last_byte_index = (self.bit + self.window_size) / 8;
+        if last_byte_index >= self.bytes.len() {
+            return None;
+        }
+
+        let mut out = 0u16;
+        for i in self.bit..self.bit + (self.window_size) {
+            out |= self.bytes.bit(i) as u16;
+        }
+        self.bit += self.window_size;
+
+        Some(out)
+    }
+}
+
+impl Bits for &[u8] {
+    fn bit(&self, bit: usize) -> u8 {
+        let byte = bit / 8;
+        let byte_bit = 7 - bit % 8;
+        (self[byte] >> byte_bit) & 1
+    }
+
+    fn iter(&self) -> BitsIter<'_> {
+        BitsIter {
+            bytes: self,
+            bit: 0,
+        }
+    }
+
+    fn iter_sliced(&self, window_size: usize) -> BitsSlicedIter<'_> {
+        BitsSlicedIter {
+            bytes: self,
+            bit: 0,
+            window_size,
+        }
+    }
+}
+
+impl Bits for Vec<u8> {
+    fn bit(&self, bit: usize) -> u8 {
+        self.as_slice().bit(bit)
+    }
+
+    fn iter(&self) -> BitsIter<'_> {
+        BitsIter {
+            bytes: self,
+            bit: 0,
+        }
+    }
+
+    fn iter_sliced(&self, window_size: usize) -> BitsSlicedIter<'_> {
+        BitsSlicedIter {
+            bytes: self,
+            bit: 0,
+            window_size,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::bit_vector::Bits;
+
+    #[test]
+    fn bits() {
+        // 00000001 00000010 00000011 00000100 00000101 00000110 ...
+        //        1        2        3        4        5        6
+        let v = vec![1u8, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];
+
+        let mut bit = 0;
+        for i in bit..7 {
+            assert_eq!(v.bit(i), 0);
+        }
+        bit += 7;
+        assert_eq!(v.bit(bit), 1);
+        bit = 8;
+        for i in bit..14 {
+            assert_eq!(v.bit(i), 0);
+        }
+        bit = 14;
+        assert_eq!(v.bit(bit), 1);
+        bit += 1;
+        assert_eq!(v.bit(bit), 0);
+
+        for bit in v.iter() {
+            eprint!("{bit}");
+        }
+        eprintln!();
+    }
+
+    #[test]
+    fn better_bits() {
+
+        // 00000001 00000010 00000011 00000100 00000101 00000110 ...
+        //        1        2        3        4        5        6
+        let v = vec![1u8, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];
+
+        let mut it = v.iter_sliced(7);
+        for i in bit..7 {
+            assert_eq!(v.bit(i), 0);
```

Comment:
This test fails to build since no variable `bit` was declared in the function; also, did you mean `assert_eq!(it.bytes.bit(i), 0)` to test the iterator?

Reply:
Ah, I thought I removed this again.
This was part of an experiment that was unnecessary. I'll delete it and what it pretends to test.